tt_metal/impl/dispatch/command_queue.cpp

```diff
@@ -393,21 +352,82 @@ const DeviceCommand EnqueueReadBufferCommand::assemble_device_command(uint32_t d
-        command.set_sharded_buffer_num_cores(num_cores);
+    return command;
+}
+
+const DeviceCommand EnqueueReadBufferCommand::assemble_device_command_sharded(uint32_t dst_address) {
+    DeviceCommand command;
+
+    uint32_t padded_page_size = align(this->buffer.page_size(), 32);
+    uint32_t num_pages = this->pages_to_read;
+    uint32_t buffer_address = this->buffer.address();
+    uint32_t dst_page_index = 0;
+
+    uint32_t num_cores = this->buffer.num_cores();
+    uint32_t shard_size = this->buffer.shard_spec().size();
+    //TODO: for now all shards are same size of pages
+    vector<uint32_t> num_pages_in_shards(num_cores, shard_size);
+    vector<uint32_t> core_id_x;
+    core_id_x.reserve(num_cores);
+    vector<uint32_t> core_id_y;
+    core_id_y.reserve(num_cores);
+    auto all_cores = this->buffer.all_cores();
+    for (const auto & core: all_cores) {
+        CoreCoord physical_core = device->worker_core_from_logical_core(core);
+        core_id_x.push_back(physical_core.x);
+        core_id_y.push_back(physical_core.y);
     }
+    command.add_buffer_transfer_instruction_sharded(
+        buffer_address,
+        dst_address,
+        num_pages,
+        padded_page_size,
+        (uint32_t)this->buffer.buffer_type(),
+        uint32_t(BufferType::SYSTEM_MEMORY),
+        this->src_page_index,
+        dst_page_index,
+        num_pages_in_shards,
+        core_id_x,
+        core_id_y
+    );
+
+    command.set_sharded_buffer_num_cores(num_cores);
+
+    return assemble_device_command_postamble(command, padded_page_size, num_pages);
+}
 
-    TT_ASSERT(padded_page_size <= consumer_cb_size, "Page is too large to fit in consumer buffer");
-    return command;
+const DeviceCommand EnqueueReadBufferCommand::assemble_device_command(uint32_t dst_address) {
+    DeviceCommand command;
+
+    uint32_t padded_page_size = align(this->buffer.page_size(), 32);
+    uint32_t num_pages = this->pages_to_read;
+    uint32_t buffer_address = this->buffer.address();
+    uint32_t dst_page_index = 0;
+
+    command.add_buffer_transfer_instruction(
+        buffer_address,
+        dst_address,
+        num_pages,
+        padded_page_size,
+        (uint32_t)this->buffer.buffer_type(),
+        uint32_t(BufferType::SYSTEM_MEMORY),
+        this->src_page_index,
+        dst_page_index);
+
+    return assemble_device_command_postamble(command, padded_page_size, num_pages);
 }
 
 void EnqueueReadBufferCommand::process() {
     uint32_t write_ptr = this->manager.get_issue_queue_write_ptr(this->command_queue_channel);
     this->read_buffer_addr = this->manager.get_completion_queue_read_ptr(this->command_queue_channel);
 
-    const auto cmd = this->assemble_device_command(this->read_buffer_addr);
+    DeviceCommand cmd;
+    if(is_sharded(this->buffer.buffer_layout())) {
+        cmd = this->assemble_device_command_sharded(this->read_buffer_addr);
+    }
+    else {
+        cmd = this->assemble_device_command(this->read_buffer_addr);
+    }
```

Comment:
as @TT-billteng suggested, I think we should make new commands for sharding, and have enqueue_read_buffer create the correct command.